pkg/backend/httpstate/client/api.go

```diff
@@ -331,7 +336,9 @@ func readBody(resp *http.Response) ([]byte, error) {
 	case "gzip":
 		logging.V(apiRequestDetailLogLevel).Infoln("decompressing gzipped response from service")
 		reader, err := gzip.NewReader(resp.Body)
-		defer contract.IgnoreClose(reader)
+		if reader != nil {
+			defer contract.IgnoreClose(reader)
+		}
```

Comment:
Is it the case that NewReader can return a nil error but also a nil reader?  I thought no. In which case - I think the typical way to write this would be to move the defer after the err check.

Reply:
That's what I assumed initially ([my initial commit](https://github.com/pulumi/pulumi/commit/bcac6cf1d772c2a0bfbba45df821d0a1a7e6c331) moved the defer IgnoreClose call to after the err check), but I didn't find any formal confirmation about this behaviour ([gzip.NewReader documentation](https://golang.org/pkg/compress/gzip/#NewReader)). I then decided to keep it here to reduce the potential side-effect that I don't know about.
If your experience confirms that in case of error the other pointer is always nil, I will move the defer call to after the error check.

Reply:
@lukehoban can you have a look here? I'd like to know if I you think I should change this part, or if it's ready to go, thanks!

Reply:
Looks fine as is - thanks!.